bug: failing test KafkaServiceStrimziKafkaRefReconcilerIT

A test failed in GitHub Actions:

## failed test

KafkaServiceStrimziKafkaRefReconcilerIT
https://github.com/kroxylicious/kroxylicious/actions/runs/25175647549/job/73806875525

looks to have failed on main since https://github.com/kroxylicious/kroxylicious/commit/1767a63938ed8b9a0e7899b464c001e7f7469847

## stack trace

```
Error:  Tests run: 9, Failures: 0, Errors: 9, Skipped: 0, Time elapsed: 20.27 s <<< FAILURE! -- in io.kroxylicious.kubernetes.operator.reconciler.kafkaservice.KafkaServiceStrimziKafkaRefReconcilerIT
Error:  io.kroxylicious.kubernetes.operator.reconciler.kafkaservice.KafkaServiceStrimziKafkaRefReconcilerIT.trustAnchorRefUsedWhenTrustStrimziCaDisabled -- Time elapsed: 2.309 s <<< ERROR!
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://192.168.49.2:8443/apis/kafka.strimzi.io/v1beta2/namespaces/kafkaservicestrimzikafkarefr-trustanchorrefusedwhentrusts-71b01/kafkas. Message: Not Found.
	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:205)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:507)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:524)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleCreate(OperationSupport.java:340)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleCreate(BaseOperation.java:766)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleCreate(BaseOperation.java:98)
	at io.fabric8.kubernetes.client.dsl.internal.CreateOnlyResourceOperation.create(CreateOnlyResourceOperation.java:42)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.create(BaseOperation.java:1167)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.create(BaseOperation.java:98)
	at io.kroxylicious.kubernetes.operator.LocallyRunningOperatorRbacHandler$1.create(LocallyRunningOperatorRbacHandler.java:210)
	at io.kroxylicious.kubernetes.operator.reconciler.kafkaservice.KafkaServiceStrimziKafkaRefReconcilerIT.trustAnchorRefUsedWhenTrustStrimziCaDisabled(KafkaServiceStrimziKafkaRefReconcilerIT.java:213)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://192.168.49.2:8443/apis/kafka.strimzi.io/v1beta2/namespaces/kafkaservicestrimzikafkarefr-trustanchorrefusedwhentrusts-71b01/kafkas. Message: Not Found.
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:642)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:622)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:582)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:549)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:141)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.utils.AsyncUtils.lambda$retryWithExponentialBackoff$3(AsyncUtils.java:91)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:844)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)

```

--- a/kroxylicious-kubernetes/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/AllReconcilersIT.java
+++ b/kroxylicious-kubernetes/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/AllReconcilersIT.java
@@ -21,7 +21,6 @@ import org.assertj.core.api.InstanceOfAssertFactories;
 import org.awaitility.core.ConditionFactory;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.condition.EnabledIf;
 import org.junit.jupiter.api.extension.RegisterExtension;
@@ -88,6 +87,7 @@ class AllReconcilersIT {
     private static final String CLUSTER_FOO_CLUSTER_IP_INGRESS = "foo-cluster-ip";
     private static final String CLUSTER_FOO_SERVICE = "foo-service";
     private static final String CLUSTER_FOO_FILTER = "foo-filter";
+    private static final String STRIMZI_TLS_LISTENER = "tls";
     private static final ConditionFactory AWAIT = await().timeout(Duration.ofSeconds(60));
 
     // the initial operator image pull can take a long time and interfere with the tests
@@ -417,94 +417,85 @@ class AllReconcilersIT {
         assertResourcesAttainCondition(AllReconcilersIT::refsResolved, myCluster, myIngress, myService);
     }
 
-    /**
-     * Nested class for tests that require Strimzi CRD
-     */
-    @Nested
-    class StrimziKafkaIntegrationTests {
-
-        private static final String STRIMZI_TLS_LISTENER = "tls";
-
-        @Test
-        void upstreamTlsFromStrimziKafkaRef() {
-            // Given
-            String kafkaName = "my-cluster";
-            // @formatter:off
-            var kafka = testActor.create(new KafkaBuilder()
-                    .withNewMetadata()
-                        .withName(kafkaName)
-                    .endMetadata()
-                    .withNewSpec()
-                        .withNewKafka()
-                            .addNewListener()
-                                .withName(STRIMZI_TLS_LISTENER)
-                                .withTls(true)
-                            .endListener()
-                        .endKafka()
-                    .endSpec()
-                    .build());
-
-            // Patch the Kafka status to simulate what the Strimzi operator would do.
-            // The Strimzi operator manages the status subresource of Kafka CRs, populating
-            // listener addresses and other runtime state. In tests, we must manually set
-            // this status since the Strimzi operator is not running.
-            testActor.patchStatus(new KafkaBuilder(kafka)
-                    .withNewStatus()
+    @Test
+    void upstreamTlsFromStrimziKafkaRef() {
+        // Given
+        String kafkaName = "my-cluster";
+        // @formatter:off
+        var kafka = testActor.create(new KafkaBuilder()
+                .withNewMetadata()
+                    .withName(kafkaName)
+                .endMetadata()
+                .withNewSpec()
+                    .withNewKafka()
                         .addNewListener()
                             .withName(STRIMZI_TLS_LISTENER)
-                            .addNewAddress()
-                                    .withHost("kafka.example.com")
-                                    .withPort(9093)
-                            .endAddress()
+                            .withTls(true)
                         .endListener()
-                    .endStatus()
-                    .build());
+                    .endKafka()
+                .endSpec()
+                .build());
 
-            testActor.create(new SecretBuilder()
-                    .withNewMetadata()
-                        .withName(kafkaName + ResourcesUtil.STRIMZI_CLUSTER_CA_CERT_SECRET_SUFFIX)
-                    .endMetadata()
-                    .addToData(ResourcesUtil.STRIMZI_CLUSTER_CA_BUNDLE, "dGVzdC1jYQ==")
-                    .build());
+        // Patch the Kafka status to simulate what the Strimzi operator would do.
+        // The Strimzi operator manages the status subresource of Kafka CRs, populating
+        // listener addresses and other runtime state. In tests, we must manually set
+        // this status since the Strimzi operator is not running.
+        testActor.patchStatus(new KafkaBuilder(kafka)
+                .withNewStatus()
+                    .addNewListener()
+                        .withName(STRIMZI_TLS_LISTENER)
+                        .addNewAddress()
+                                .withHost("kafka.example.com")
+                                .withPort(9093)
+                        .endAddress()
+                    .endListener()
+                .endStatus()
+                .build());
 
-            var myService = editableStrimziService(CLUSTER_FOO_SERVICE, kafkaName, STRIMZI_TLS_LISTENER).build();
-            var myProxy = editableProxy(PROXY_A).build();
-            var myIngress = editableIngress(CLUSTER_FOO_CLUSTER_IP_INGRESS, myProxy)
-                    .editOrNewSpec()
-                        .withNewClusterIP()
-                            .withProtocol(Protocol.TCP)
-                        .endClusterIP()
-                    .endSpec()
-                    .build();
-            // @formatter:on
+        testActor.create(new SecretBuilder()
+                .withNewMetadata()
+                    .withName(kafkaName + ResourcesUtil.STRIMZI_CLUSTER_CA_CERT_SECRET_SUFFIX)
+                .endMetadata()
+                .addToData(ResourcesUtil.STRIMZI_CLUSTER_CA_BUNDLE, "dGVzdC1jYQ==")
+                .build());
 
-            var myCluster = editableVirtualCluster(CLUSTER_FOO, myProxy, myService, List.of(myIngress), List.of()).build();
+        var myService = editableStrimziService(CLUSTER_FOO_SERVICE, kafkaName, STRIMZI_TLS_LISTENER).build();
+        var myProxy = editableProxy(PROXY_A).build();
+        var myIngress = editableIngress(CLUSTER_FOO_CLUSTER_IP_INGRESS, myProxy)
+                .editOrNewSpec()
+                    .withNewClusterIP()
+                        .withProtocol(Protocol.TCP)
+                    .endClusterIP()
+                .endSpec()
+                .build();
+        // @formatter:on
 
-            // When
-            createAll(myProxy, myCluster, myIngress, myService);
+        var myCluster = editableVirtualCluster(CLUSTER_FOO, myProxy, myService, List.of(myIngress), List.of()).build();
 
-            // Then
-            assertResourcesAttainCondition(AllReconcilersIT::resourceReady, myProxy);
-            assertResourcesAttainCondition(AllReconcilersIT::refsResolved, myCluster, myIngress, myService);
-        }
+        // When
+        createAll(myProxy, myCluster, myIngress, myService);
 
-        private static KafkaServiceBuilder editableStrimziService(String name, String kafkaName, String listenerName) {
-            // @formatter:off
-            return new KafkaServiceBuilder()
-                    .withNewMetadata()
-                        .withName(name)
-                    .endMetadata()
-                    .withNewSpec()
-                        .withNewStrimziKafkaRef()
-                            .withListenerName(listenerName)
-                            .withTrustStrimziCaCertificate(true)
-                            .withNewRef()
-                                .withName(kafkaName)
-                            .endRef()
-                        .endStrimziKafkaRef()
-                    .endSpec();
-            // @formatter:on
-        }
+        // Then
+        assertResourcesAttainCondition(AllReconcilersIT::resourceReady, myProxy);
+        assertResourcesAttainCondition(AllReconcilersIT::refsResolved, myCluster, myIngress, myService);
+    }
+
+    private static KafkaServiceBuilder editableStrimziService(String name, String kafkaName, String listenerName) {
+        // @formatter:off
+        return new KafkaServiceBuilder()
+                .withNewMetadata()
+                    .withName(name)
+                .endMetadata()
+                .withNewSpec()
+                    .withNewStrimziKafkaRef()
+                        .withListenerName(listenerName)
+                        .withTrustStrimziCaCertificate(true)
+                        .withNewRef()
+                            .withName(kafkaName)
+                        .endRef()
+                    .endStrimziKafkaRef()
+                .endSpec();
+        // @formatter:on
     }
 
     private void createAll(HasMetadata... resources) {

--- a/kroxylicious-kubernetes/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/AllReconcilersIT.java
+++ b/kroxylicious-kubernetes/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/AllReconcilersIT.java
@@ -19,7 +19,6 @@ import java.util.stream.Stream;
 
 import org.assertj.core.api.InstanceOfAssertFactories;
 import org.awaitility.core.ConditionFactory;
-import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Nested;
@@ -36,8 +35,6 @@ import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
 import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.SecretBuilder;
 import io.fabric8.kubernetes.client.CustomResource;
-import io.fabric8.kubernetes.client.KubernetesClient;
-import io.fabric8.kubernetes.client.dsl.Updatable;
 import io.javaoperatorsdk.operator.junit.LocallyRunOperatorExtension;
 import io.strimzi.api.kafka.Crds;
 import io.strimzi.api.kafka.model.kafka.KafkaBuilder;
@@ -113,6 +110,7 @@ class AllReconcilersIT {
             .withKubernetesClient(rbacHandler.operatorClient())
             .waitForNamespaceDeletion(false)
             .withConfigurationService(x -> x.withCloseClientOnStop(false))
+            .withAdditionalCustomResourceDefinition(Crds.kafka())
             .build();
     private final TestActor testActor = rbacHandler.testActor(extension);
 
@@ -426,26 +424,6 @@ class AllReconcilersIT {
     class StrimziKafkaIntegrationTests {
 
         private static final String STRIMZI_TLS_LISTENER = "tls";
-
-        @BeforeAll
-        static void setUpStrimziCrd() {
-            try (KubernetesClient client = OperatorTestUtils.kubeClient()) {
-                client.apiextensions().v1().customResourceDefinitions().resource(Crds.kafka()).createOr(Updatable::update);
-            }
-            catch (Exception e) {
-                throw new RuntimeException("Failed to set up Strimzi CRD", e);
-            }
-        }
-
-        @AfterAll
-        static void tearDownStrimziCrd() {
-            try (KubernetesClient client = OperatorTestUtils.kubeClient()) {
-                client.apiextensions().v1().customResourceDefinitions().resource(Crds.kafka()).delete();
-            }
-            catch (Exception e) {
-                LOGGER.atWarn().setCause(e).log("Failed to clean up Strimzi CRD");
-            }
-        }
 
         @Test
         void upstreamTlsFromStrimziKafkaRef() {

--- a/kroxylicious-kubernetes/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaservice/KafkaServiceStrimziKafkaRefReconcilerIT.java
+++ b/kroxylicious-kubernetes/kroxylicious-operator/src/test/java/io/kroxylicious/kubernetes/operator/reconciler/kafkaservice/KafkaServiceStrimziKafkaRefReconcilerIT.java
@@ -11,9 +11,7 @@ import java.time.Duration;
 
 import org.assertj.core.api.Assertions;
 import org.awaitility.core.ConditionFactory;
-import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.condition.EnabledIf;
 import org.junit.jupiter.api.extension.RegisterExtension;
@@ -24,8 +22,6 @@ import io.fabric8.kubernetes.api.model.ConfigMap;
 import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
 import io.fabric8.kubernetes.api.model.Secret;
 import io.fabric8.kubernetes.api.model.SecretBuilder;
-import io.fabric8.kubernetes.client.KubernetesClient;
-import io.fabric8.kubernetes.client.dsl.Updatable;
 import io.javaoperatorsdk.operator.junit.LocallyRunOperatorExtension;
 import io.strimzi.api.kafka.Crds;
 import io.strimzi.api.kafka.model.kafka.Kafka;
@@ -42,7 +38,6 @@ import io.kroxylicious.kubernetes.api.v1alpha1.KafkaServiceStatus;
 import io.kroxylicious.kubernetes.api.v1alpha1.kafkaservicestatus.Tls;
 import io.kroxylicious.kubernetes.operator.Annotations;
 import io.kroxylicious.kubernetes.operator.LocallyRunningOperatorRbacHandler;
-import io.kroxylicious.kubernetes.operator.OperatorTestUtils;
 import io.kroxylicious.kubernetes.operator.ResourcesUtil;
 import io.kroxylicious.kubernetes.operator.TestFiles;
 import io.kroxylicious.kubernetes.operator.assertj.OperatorAssertions;
@@ -77,29 +72,8 @@ class KafkaServiceStrimziKafkaRefReconcilerIT {
             .withKubernetesClient(rbacHandler.operatorClient())
             .waitForNamespaceDeletion(false)
             .withConfigurationService(x -> x.withCloseClientOnStop(false))
+            .withAdditionalCustomResourceDefinition(Crds.kafka())
             .build();
-
-    @BeforeAll
-    static void beforeAll() {
-        // note that we could not find a nice way to do this via the LocallyRunOperatorExtension. I tried serializing the CRD to
-        // a temp file and using `withAdditionalCRD(path)` but it didn't load those CRDs before initializing the reconciler.
-        try (KubernetesClient client = OperatorTestUtils.kubeClient()) {
-            client.apiextensions().v1().customResourceDefinitions().resource(Crds.kafka()).createOr(Updatable::update);
-        }
-        catch (Exception e) {
-            throw new RuntimeException(e);
-        }
-    }
-
-    @AfterAll
-    static void afterAll() {
-        try (KubernetesClient client = OperatorTestUtils.kubeClient()) {
-            client.apiextensions().v1().customResourceDefinitions().resource(Crds.kafka()).delete();
-        }
-        catch (Exception e) {
-            throw new RuntimeException(e);
-        }
-    }
 
     private final LocallyRunningOperatorRbacHandler.TestActor testActor = rbacHandler.testActor(extension);
 

